pkg/auth/dependency/mfa/pq/store.go

```diff
@@ -0,0 +1,162 @@
+package pq
+
+import (
+	"github.com/lib/pq"
+
+	"github.com/skygeario/skygear-server/pkg/auth/dependency/mfa"
+	coreAuth "github.com/skygeario/skygear-server/pkg/core/auth"
+	"github.com/skygeario/skygear-server/pkg/core/config"
+	"github.com/skygeario/skygear-server/pkg/core/db"
+	"github.com/skygeario/skygear-server/pkg/core/time"
+	"github.com/skygeario/skygear-server/pkg/core/uuid"
+)
+
+type storeImpl struct {
+	mfaConfig    config.MFAConfiguration
+	sqlBuilder   db.SQLBuilder
+	sqlExecutor  db.SQLExecutor
+	timeProvider time.Provider
+}
+
+func NewStore(
+	mfaConfig config.MFAConfiguration,
+	sqlBuilder db.SQLBuilder,
+	sqlExecutor db.SQLExecutor,
+	timeProvider time.Provider,
+) mfa.Store {
+	return &storeImpl{
+		mfaConfig:    mfaConfig,
+		sqlBuilder:   sqlBuilder,
+		sqlExecutor:  sqlExecutor,
+		timeProvider: timeProvider,
+	}
+}
+
+func (s *storeImpl) GetRecoveryCode(userID string) (output []mfa.RecoveryCodeAuthenticator, err error) {
+	builder := s.sqlBuilder.Tenant().
+		Select(
+			"a.id",
+			"a.user_id",
+			"a.type",
+			"arc.code",
+			"arc.created_at",
+			"arc.consumed",
+		).
+		From(s.sqlBuilder.FullTableName("authenticator"), "a").
+		Join(
+			s.sqlBuilder.FullTableName("authenticator_recovery_code"),
+			"arc",
+			"a.id = arc.id",
+		).
+		Where("a.user_id = ?", userID)
+	rows, err := s.sqlExecutor.QueryWith(builder)
+	if err != nil {
+		return
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var a mfa.RecoveryCodeAuthenticator
+		err = rows.Scan(
+			&a.ID,
+			&a.UserID,
+			&a.Type,
+			&a.Code,
+			&a.CreatedAt,
+			&a.Consumed,
+		)
+		if err != nil {
+			return
+		}
+		output = append(output, a)
+	}
+
+	return
+}
+
+func (s *storeImpl) GenerateRecoveryCode(userID string) ([]mfa.RecoveryCodeAuthenticator, error) {
+	old, err := s.GetRecoveryCode(userID)
+	if err != nil {
+		return nil, err
+	}
+	var ids []string
+	for _, a := range old {
+		ids = append(ids, a.ID)
+	}
+
+	if len(ids) > 0 {
+		q1 := s.sqlBuilder.Tenant().
```

Comment:
Did not think of `DELETE CASCADE` when I wrote this. But it is quite dangerous operation so I prefer we keep the origin approach.

Reply:
I think it's better to use `ON DELETE CASCADE` than deleting the related records manually. It would be safer to maintain referential integrity through relational mechanism than maintain it manually.

Reply:
I have an opposite opinion. Deleting manually states the relation in the code. If we add additional relation in future and cascading delete does not apply, then cascading delete is dangerous. Manual delete will save us because the foreign key constraint will trigger error.

Reply:
I agree that cascade should not be used if the foreign key is used for ownership (e.g. app created_by -> user id). However, this scenario is inheritance/composition (i.e. user contains some recovery code authenticator; recovery code authenticator is composed of single recovery code), where the referencing row is logically the same entity as referenced row. I think using cascade delete is appropriate in this scenario.

Reply:
I now hide the delete cascading delete logic in the store implementation so provider is unaware of it.